### DISCUSSION TITLE
Allowed different entity with same name

### DISCUSSION THREAD

**Discussion by wenerme** [mikro-orm/mikro-orm#5614] [Ideas]
**Is your feature request related to a problem? Please describe.**

Some entity are transient, see https://github.com/mikro-orm/mikro-orm/discussions/5371 , I hope I can have multi eneity with same Name.

**Describe the solution you'd like**

Provide more way to help `MetadataStorage.getMetadata` generate key

1. `@Entity` support a `tag` or `name` option , tag will append to className, name will replace className
2. Use a WeakMap to track automatic global id gen
3. accept a custome way to generate entityName  ?
4. check a static method in Entity, e.g. `XEntity.getEntityName`
5. uniform the way to get entity name, e.g. `Utils.getEntityName`

Maybe only do this for abstract Entity or match the custom condition ? 

But the problem is the `Decorator` run's very early stage, this makes the the config/patch very hard, if we can delay or lazy the metadata eval, will make the postprocess easier.

**Describe alternatives you've considered**

 

**Additional context**

Here is a monkey patch works for https://github.com/mikro-orm/mikro-orm/discussions/5371#discussioncomment-9575385

```ts
import { Dictionary, EntityMetadata, MetadataStorage, Utils } from '@mikro-orm/core';

export function patchMikroORMMetadataStorage() {
  if (patchMikroORMMetadataStorage.origin) {
    return;
  }
  patchMikroORMMetadataStorage.origin = MetadataStorage.getMetadataFromDecorator;

  let _GlobalEntityId = 0;
  let _map = new WeakMap();
  MetadataStorage.getMetadataFromDecorator = <T = any>(target: T & Dictionary): EntityMetadata<T> => {
    let id = _map.get(target);
    if (!id) {
      id = _GlobalEntityId++;
      _map.set(target, id);
    }
    let name = target.name;
    // restrict to mixin entity
    if (!name.endsWith('__') && name.endsWith('MixinEntity')) {
      name = `${target.name}__${id}__`;
    }
    const path = Utils.lookupPathFromDecorator(name);
    const meta = MetadataStorage.getMetadata(name, path);
    Object.defineProperty(target, '__path', { value: path, writable: true });
    Object.defineProperty(target, 'name', { value: name, writable: true });
    return meta;
  };
}

patchMikroORMMetadataStorage.origin = null as typeof MetadataStorage.getMetadataFromDecorator | null;
```

must run in very early stage by lazy import the main

```ts
import { loadEnvs } from '@wener/nestjs';
import { App } from '@wener/nestjs/app';
import { patchMikroORMMetadataStorage } from './patchMikroORMMetadataStorage';

patchMikroORMMetadataStorage();
process.env.APP_NAME = 'wener';
process.env.APP_COMPONENT = 'api-server';
await loadEnvs({ name: App.service });
const { runApiServer } = await import('@/modules/api/runApiServer');
await runApiServer();
```

**Comment by B4nan** [comment]
I am not interested in doing this, to me this is a wontfix, you cant have entities with the same name. Your proposal solves only a small part of the problem, there are still a lot of other places that expect the entity name to be unique and will break.

**Reply by wenerme** [reply]
How about allowed this behavior for only small part of Entity ? I think EntityName should be unique across major Entity, but js allowed create dynamic classes, when work with the dynamic created classes, the dynamic created entity should be allowed have same name.
